src/cdp/cdp.zig

```diff
@@ -0,0 +1,316 @@
+    const options: std.json.StringifyOptions = .{ .emit_null_optional_fields = false };
+
+    try std.json.stringify(res, options, out.writer());
+    const ret = try alloc.alloc(u8, out.items.len);
+    @memcpy(ret, out.items);
+    return ret;
+}
+
+const resultNull = "{{\"id\": {d}, \"result\": {{}}}}";
+const resultNullSession = "{{\"id\": {d}, \"result\": {{}}, \"sessionId\": \"{s}\"}}";
+
+// caller owns the slice returned
+pub fn result(
+    alloc: std.mem.Allocator,
+    id: u16,
+    comptime T: ?type,
+    res: anytype,
+    sessionID: ?[]const u8,
+) ![]const u8 {
+    if (T == null) {
+        // No need to stringify a custom JSON msg, just use string templates
+        if (sessionID) |sID| {
+            return try std.fmt.allocPrint(alloc, resultNullSession, .{ id, sID });
+        }
+        return try std.fmt.allocPrint(alloc, resultNull, .{id});
+    }
+
+    const Resp = struct {
+        id: u16,
+        result: T.?,
+        sessionId: ?[]const u8,
+    };
+    const resp = Resp{ .id = id, .result = res, .sessionId = sessionID };
+
+    return stringify(alloc, resp);
+}
+
+pub fn sendEvent(
+    alloc: std.mem.Allocator,
+    ctx: *Ctx,
+    name: []const u8,
+    comptime T: type,
+    params: T,
+    sessionID: ?[]const u8,
+) !void {
+    const Resp = struct {
+        method: []const u8,
+        params: T,
+        sessionId: ?[]const u8,
+    };
+    const resp = Resp{ .method = name, .params = params, .sessionId = sessionID };
+
+    const event_msg = try stringify(alloc, resp);
+    std.log.debug("event {s}", .{event_msg});
+    try server.sendSync(ctx, event_msg);
+}
+
+fn getParams(
+    alloc: std.mem.Allocator,
+    comptime T: type,
+    scanner: *std.json.Scanner,
+    key: []const u8,
+) !?T {
+
+    // check key is "params"
+    if (!std.mem.eql(u8, "params", key)) return null;
+
+    // skip "params" if not requested
+    if (T == void) {
+        var finished: usize = 0;
+        while (true) {
+            switch (try scanner.next()) {
+                .object_begin => finished += 1,
+                .object_end => finished -= 1,
+                else => continue,
+            }
+            if (finished == 0) break;
+        }
+        return void{};
+    }
+
+    // parse "params"
+    const options = std.json.ParseOptions{
+        .max_value_len = scanner.input.len,
+        .allocate = .alloc_if_needed,
+    };
+    return try std.json.innerParse(T, alloc, scanner, options);
+}
+
+fn getId(scanner: *std.json.Scanner, key: []const u8) !?u16 {
+
+    // check key is "id"
+    if (!std.mem.eql(u8, "id", key)) return null;
+
+    // parse "id"
+    return try std.fmt.parseUnsigned(u16, (try scanner.next()).number, 10);
+}
+
+fn getSessionId(scanner: *std.json.Scanner, key: []const u8) !?[]const u8 {
+
+    // check key is "sessionId"
+    if (!std.mem.eql(u8, "sessionId", key)) return null;
+
+    // parse "sessionId"
+    return (try scanner.next()).string;
+}
+
+pub fn getMsg(
```

Comment:
The caller must free the the params isn't it?
It doesn't seem to be easy since params are eventually allocated :/

Reply:
The allocator is needed to call `std.json.innerParse` on params (when params_T is provided).
In my understanding this function only allocates temporary for token and free them before returning.
This function uses things like `@memcopy` from the original scanner to avoid allocating the params itself.